### PR TITLE
[jit] Add list inequality operator

### PR DESCRIPTION
**Pull request by driazati**
This PR adds `aten::neq` for list inequality comparisons and converts
`nll_loss` to weak script

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1607,6 +1608,8 @@ def nll_loss(input, target, weight=None, size_average=None, ignore_index=-100,
             out = torch._C._nn.nll_loss2d(
                 input, target, weight, reduction_enum, ignore_index)
             ret = out.view(out_size)
+    else:
```
Comment (suo) [inline torch/nn/functional.py]: why is this necessary?

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -699,32 +699,57 @@ Operation listEq(const Node* node) {
   };
 }
 
+template <typename T>
+Operation listNe(const Node* node) {
+  return [=](Stack& stack) {
+    T a;
+    T b;
+    pop(stack, a, b);
+    push(stack, a->elements() == b->elements() ? true : false);
```
Comment (apaszke) [inline torch/csrc/jit/register_prim_ops.cpp]: This doesn't conform to Python semantics, which uses lexicographic ordering on lists.

**Review by apaszke — DISMISSED** [review]

File: torch/nn/functional.py (hunk 1 above)
Comment (eellison) [reply torch/nn/functional.py]: In this case it's unnecessary because the `elif dim == 3 or dim > 4:` condition can be made into an else statement, since all other values of dim have already been handled. 

But in general we don't yet have any logic for exceptions in control-flow:
```
if cond:
    ret = 1
else:
    throw Exception()i
```

Since ret is not set in the else branch, the if node does not not emit it, even thought it should.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: How exactly does this not conform? `Tuples and lists are compared lexicographically using comparison of corresponding elements. This means that to compare equal, each element must compare equal and the two sequences must be of the same type and have the same length.`

File: torch/nn/functional.py [hunk 3]
```diff
@@ -1607,6 +1608,8 @@ def nll_loss(input, target, weight=None, size_average=None, ignore_index=-100,
...
+        ret = input  # TODO: remove when jit supports control flow analysis
```
Comment (eellison) [inline torch/nn/functional.py]: @driazati could you make the last elif branch into an else branch and delete this? this code will never be reached

File: torch/nn/functional.py [hunk 4]
```diff
@@ -1590,7 +1591,8 @@ def nll_loss(input, target, weight=None, size_average=None, ignore_index=-100,
         ret = torch._C._nn.nll_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index)
     elif dim == 4:
         ret = torch._C._nn.nll_loss2d(input, target, weight, _Reduction.get_enum(reduction), ignore_index)
-    elif dim == 3 or dim > 4:
+    else:
+        # dim == 3 or dim > 4
```
Comment (wanchaol) [inline torch/nn/functional.py]: maybe remove comment?

**Review by wanchaol — APPROVED** [review]
LG

File: torch/csrc/jit/register_prim_ops.cpp [hunk 5]
```diff
@@ -699,32 +699,57 @@ Operation listEq(const Node* node) {
...
+  }
+
+  for (size_t i = 0; i < a->elements().size(); ++i) {
+    const auto& a_element = a->elements()[i];
+    const auto& b_element = b->elements()[i];
+    // This preserves Python's semantics, which uses eq() to compare two
+    // elements, then passes the result to bool().
+    // see: https://docs.python.org/3.4/reference/datamodel.html#object.__ge__
+    const auto cmp_result = a_element.eq(b_element);
+    if (!cmp_result.is_nonzero()) {
```
Comment (apaszke) [inline torch/csrc/jit/register_prim_ops.cpp]: That doesn't really conform to Python semantics, which will throw if the tensors weren't scalar tensors.
